packages/dds/tree/docs/Shared Tree Philosophy.md

```diff
@@ -0,0 +1,254 @@
+For example the fields of object nodes are presented as JavaScript enumerable own properties.
+Additionally `simple-tree` restricts where different kinds of fields can be used to help with this.
+For example `sequence` fields are only permitted via `array` nodes, which present the node and its sequence field like an array.
+To assist with this `simple-tree` leverages view schema which allow the application to provide guidance on how to construct the API for each type of node.
+
+# Constraints
+
+When users edit documents, those high level semantic operations need to be encoded in a way that supports merges.
+These merges need to balance two competing requirements:
+
+1. The merges apply both sets of edits.
+2. Applications can easily ensure the invariants they require will be maintained across merges.
+
+The second requirement can always be met by discarding one of the edits, while the first can be met by always adjusting the second editing into something that can be applied.
+This places these two requirements in conflict with each-other.
+
+To keep developers from falling into the common pitfalls of merging violating application invariants, the editing APIs are designed to ensure that which invariants are maintained by default is clear.
+When these invariants are insufficient, the application author will always have the option to apply a constraint which will ensure than if the edit could result in a constraint violation due to a merge, it will be rejected instead of merged.
+The most important of these constraints, the one which will always be enough, is to reject the edit if any other concurrent edit got sequenced before it.
+Having this option available ensues the application author can always author their editing logic with confidence it won't behave unexpectedly due to concurrency, which is key to ensuring they can be productive.
+
+This general purpose no-concurrency constraint is unfortunately not great for the "The merges apply both sets of edits" requirement.
+This can be addressed by providing more specific constraints, and designing the editing APIs so constraints are needed in less cases.
+For example allowing a constraint that only rejects a transaction if a field or subtree was concurrently edited, or providing a field kind with higher level editing operations like a set or counter with appropriate editing methods.
+
+This design pattern ensures there is always a way for the application author to ensure correctness, and by a little doing more work to be less conservative with the constraints, the application can improve the merge granularity when needed.
+
+Additionally, even higher merge fidelity can be achieved by adding new field kinds as application needs for them arise, and this work can be shared among all users of SharedTree.
+
+Note that the current implementation does not yet provide access to constraints, but this is planned for future versions.
```

Comment:
Is that still true? I thought @noencke was adding that to the API recently.

Reply:
They're not checked in, unfortunately. We reprioritized some other work instead for the upcoming deadlines.

Reply:
Feel free to resolve without further changes